CardiorespiratoryFitness/CardiorespiratoryFitness/iOS/CRFHeartRateSampleProcessor.swift

```diff
@@ -59,16 +249,223 @@ internal class CRFHeartRateSampleProcessor {
+                confidence = peak_pos.map({ y[$0] - y_min }).mean() / (x.max()! - x.min()!)
+            }
+            else {
+                hr = hr_initial_guess
+                confidence = y_max / x.max()!
+            }
+        }
+        // # Get into this loop if no earlier peak was picked up during getAliasingPeakLocation
+        else if aliasedPeak.later.count > 0, aliasedPeak.later.reduce(true, { $0 && (y[$1] > 0.7 * y_max) }) {
+            hr = hr_initial_guess
+            confidence = y_max / x.max()!
+        }
+        else {
+            return (0, 0)
+        }
+        
+        return (hr, confidence)
+    }
+    
+    func calculateMinLag(samplingRate: Double) -> Int {
+        return Int(round(60 * samplingRate / CRFMaxHeartRate))
+    }
+    
+    func calculateMaxLag(samplingRate: Double) -> Int {
+        return Int(round(60 * samplingRate / CRFMinHeartRate))
+    }
+    
+    func preprocessSamples(_ input: [Double], samplingRate: Double) -> (x: [Double], y: [Double], minLag: Int, maxLag: Int)? {
+        //    max_lag = round(60 * sampling_rate / min_hr) # 4/3 fs is 45BPM
+        let maxLag = calculateMaxLag(samplingRate: samplingRate)
+        //    min_lag = round(60 * sampling_rate / max_hr) # 1/3.5 fs is 210BPM
+        let minLag = calculateMinLag(samplingRate: samplingRate)
+        //    x <- stats::acf(x, lag.max = max_lag, plot = F)$acf
+        let x = autocorrelation(input, lagMax: maxLag).offsetR()
+        
+        // Check that the sampling rate is valid and the max/min are within range.
+        let roundedSamplingRate = Int(round(samplingRate))
+        guard isValidSamplingRate(roundedSamplingRate),
+            maxLag < x.count, minLag < maxLag, minLag > 0
+            else {
+                return nil
+        }
+        
+        //    y <- 0 * x
+        //    y[seq(min_lag, max_lag)] <- x[seq(min_lag, max_lag)]
+        var y = Array(repeating: Double(0), count: x.count)
+        for ii in minLag...maxLag {
+            y[ii] = x[ii]
+        }
+        
+        return (x, y, minLag, maxLag)
+    }
+    
+    func getBounds(y: [Double], samplingRate: Double) -> (y_max: Double, y_max_pos: Int, y_min: Double, hr_initial_guess: Double) {
+        //    y_max_pos <- which.max(y)
+        //    y_max <- max(y)
+        //    y_min <- min(y)
+        let (y_max, y_max_pos) = y.seekMax()
+        let y_min = y.min()!
+        
+        //    hr_initial_guess <- 60 * sampling_rate / (y_max_pos - 1)
+        let hr_initial_guess = (60.0 * samplingRate) / Double(y_max_pos - 1)
+        
+        return (y_max, y_max_pos, y_min, hr_initial_guess)
+    }
+    
+    
+    func getAliasingPeakLocation(hr: Double, actualLag: Int, samplingRate: Double, minLag: Int, maxLag: Int) -> (nPeaks: Int, earlier:[Int], later: [Int])? {
+        //    # The following ranges are only valid if the minimum hr is 45bpm and
+        //    # maximum hr is less than 240bpm, since for the acf of the ideal hr signal
```

Comment:
CRFMaxHeartRate is 210...?

Reply:
That can be changed to 225. 

I just put in 210 because getting a heart rate above 210 is bit rare. 

The problem with putting a higher estimate on the maxHR is, inherently the HR's higher than 210 or so will have a larger value in the ACF (because the way ACF is calculated - lower lags have higher values), and I did not want them to interfere with the estimates. So I put a tighter bound on the maxHR